Support serialization as float in TimeDelta field

marcosatti opened the pull request:
Add functionality to (de)serialize to float values as a representation of the total seconds. 

In our use case, we are trying to specify a period which is a combination of seconds, milliseconds and microseconds, which is easier to represent as a single floating point value rather than specifying the smallest unit (microseconds). We have currently been serializing to a `fields.Float()` using `.total_seconds()` and the equivalent to deserialize.

The feature does not promise absolute precision, which is stated in the docs (same restrictions as floating point values).

This does not change the current way `TimeDelta` is used.

marcosatti committed:
Add total seconds float support to the TimeDelta field

Able to (de)serialise to float values as a representation of the total seconds.

lafrech commented on tests/test_serialization.py:
```diff
@@ -721,6 +733,19 @@ def test_timedelta_field(self, user):
+        )
+        field = fields.TimeDelta(fields.TimeDelta.TOTAL_SECONDS)
+        # Test for reasonable approximate equality. No guarantees are made about accuracy.
+        assert abs(field.serialize("d10", user) - 1130751.010742) < 0.1
```
I'd use math.isclose for this.

lafrech reviewed (commented):
Thanks for the contribution.

I understand your use case.

I'm reluctant to integrate this as is, though. At this point, we might as well offer int and float for any precision. This would have the benefit of symmetry, the possibilities would be the cartesian product of {all precisions} x {int, float}. And one could serialize to a floating point number of days.

You'd just have to add a new parameter expressing int or float (int by default).

Just my thoughts, I could be convinced otherwise and maybe other maintainers will chime in with a different view.

marcosatti commented:
Having a `serde_type` parameter does sound like a more complete approach to the problem - I'd be ok to add that in. I'll wait for more discussion first.

marcosatti committed:
Add support for all precisions when (de)serializing TimeDelta's as floats

marcosatti commented:
Went ahead and updated anyway to support all precisions through use of the `serde_type` parameter.

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1479,20 +1500,36 @@ def __init__(self, precision: str = SECONDS, **kwargs):
+            float,
+        )
+
+        if serde_type not in serde_types:
```
Nitpick: Either I'd write `if serde_type not in (int, float)` or I'd set `SERDE_TYPES` as class attribute.

I use a class attribute when I expect subclassing. In this case, I don't see much use for subclassing, so both suit me.

marcosatti replied:
Cleaned this up to be more inline.

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1461,7 +1480,9 @@ class TimeDelta(Field):
 
-    def __init__(self, precision: str = SECONDS, **kwargs):
+    def __init__(
+        self, precision: str = SECONDS, serde_type: type[int | float] = int, **kwargs
```
I don't really like the name. I suppose you mean serialization/deserialization type. Although it is involved in both serialization and deserialization, to me the name implies the destination type, so it is the serialization type while the deserialization type is `dt.timedelta`.

Therefore, I'd call it `serialization_type` or `ser_type`. Those doesn't sound so nice, so I'm open to suggestions.

marcosatti replied:
Went with `serialization_type`... Maybe `primitive_type` could be used but I feel the former describes it better.

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1479,20 +1500,36 @@ def __init__(self, precision: str = SECONDS, **kwargs):
+
+        if serde_type not in serde_types:
+            msg = f"The (de)serialization must be one of {serde_types}"
+            raise ValueError(msg)
```
I don't see a test for this, but I'm not sure it is required.

marcosatti replied:
I put one into `test_serialization.py` although maybe its not the correct place for it...

lafrech replied:
Oh, right. Didn't see it.

lafrech reviewed (commented):
Nice, thanks!

Just minor comments.

marcosatti committed:
Fix small issues with TimeDelta serialization type

- Implement a test for exception being raised on invalid type parameter.
- Rename serde_type to serialization_type.
- Condense code.

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1434,17 +1434,36 @@ def _make_object_from_format(value, data_format):
         Add `precision` parameter.
+    .. versionchanged:: 3.17.0
+        Allow (de)serialization to `float` through use of a new `serde_type` parameter.
+        `int` is the default to retain previous behaviour.
```
> Allow (de)serialization to `float` through use of a new `serde_type` parameter.

Wrong param name.

lafrech commented on src/marshmallow/fields.py:
```diff
@@ -1434,17 +1434,36 @@ def _make_object_from_format(value, data_format):
+    Microseconds is the smallest unit able to be represented wholly.
+    Smaller units than microseconds cannot be properly deserialized and will result
+    in the truncation of any fractional data. This is only possible when using `float`.
+    For example, 1.12345 interpreted as microseconds will result in `timedelta(microseconds=1)`.
```
I'm not sure about the meaning of "This is only possible when using `float`." I suppose you mean only float is affected because there is no fractional part in int. But I found the sentence confusing. Just removing it might make things clearer. The example clarifies things anyway.

Or perhaps rephrase and state that whatever the precision chosen, the fractional part when using float will always be truncated to microseconds.

lafrech reviewed (approved):
LGTM. Just a few comments.

I'll wait a few days in case other maintainers want to review as well.

marcosatti committed:
Fix small issues with TimeDelta documentation

marcosatti commented:
Fixed up remaining documentation issues.
